File: src/support/GBRandomState.java
/*******************************************************************************
 * Copyright (c) 2002-2013 (c) Devon and Warren Schudy
 * Copyright (c) 2014  Devon and Warren Schudy, Mike Anderson
 *******************************************************************************/
package support;

import java.awt.Color;

//Modified to use java's built-in Random class

public class GBRandomState {
	int seed;
	java.util.Random rnd;

	// convenient global generator
	public static GBRandomState gRandoms = new GBRandomState();

	public GBRandomState() {
		rnd = new java.util.Random();
		seed = rnd.nextInt();
	}

	public GBRandomState(int newseed) {
		seed = newseed;
	}

	public int Generateint() {
		return rnd.nextInt();
	}

	public short GenerateShort() {
		return (short) (Generateint() >> 16);
	}

	public int intInRange(int min, int max) {
		if (min == max)
			return min;
		else if (min > max)
			// With this limitation we can't generate negative numbers, but
			// changing it might change robot behavior
			return intInRange(max, min);
		int result = rnd.nextInt(max - min + 1) + min;
		return result;
	}

	public double InRange(double min, double max) {
		return rnd.nextDouble() * (max - min) + min;
	}

	public float FloatInRange(float min, float max) {
		return rnd.nextFloat() * (max - min) + min;
	}

	public double Angle() {
		return InRange(GBMath.kEpsilon - GBMath.kPi, GBMath.kPi);
	}

	public FinePoint Vector(double maxLength) {
		return FinePoint.makePolar(InRange(0, maxLength), Angle());
	}

	public Color Color() {
		return new Color(FloatInRange(0, 1), FloatInRange(0, 1), FloatInRange(
				0, 1));
	}

	// TODO: This was marked as broken. Fixed syntax but was there more to it?
	public Color ColorNear(Color color, float dist) {
		float r = FloatInRange(color.getRed() / 255f - dist, color.getRed()
				/ 255f + dist);
		float g = FloatInRange(color.getGreen() / 255f - dist, color.getGreen()
				/ 255f + dist);
		float b = FloatInRange(color.getBlue() / 255f - dist, color.getBlue()
				/ 255f + dist);
		return new Color(r, g, b);
	}

	public boolean bool(double probability) {
		return InRange(0, 1 - GBMath.kEpsilon) < probability;
	}

	public boolean bool(int num, int denom) {
		return intInRange(0, denom - 1) < num;
	}

	public int GetSeed() {
		return seed;
	}

	public void SetSeed(int newSeed) {
		seed = newSeed;
	}
}

File: grobots_cpp/support/GBRandomState.cpp
// GBRandomState.cpp
// Grobots (c) 2002-2004 Devon and Warren Schudy
// Distributed under the GNU General Public License.

#include "GBRandomState.h"
#include "GBErrors.h"
#include "GBStringUtilities.h"
#include "GBColor.h"
#include <time.h>


// convenient global generator
GBRandomState gRandoms;


GBRandomState::GBRandomState()
	: seed((long)time(0))
{}

GBRandomState::GBRandomState(const long newseed)
	: seed(newseed)
{}

long GBRandomState::GenerateLong() {
	seed = seed * 0734652105 + 662049451; // constants as recommended by Knuth
	return seed;
}

short GBRandomState::GenerateShort() {
	return (short)(GenerateLong() >> 16);
}

long GBRandomState::LongInRange(const long min, const long max) {
	if ( min == max )
		return min;
	else if ( min > max )
		return LongInRange(max, min);
	unsigned long range = max - min + 1;
	long result = min + (((GBLongLong)(unsigned long)GenerateLong() * range) >> 32);
	if ( result < min || result > max ) {
		NonfatalError(string("GBRandomState::LongInRange(") + ToString(min) + ", "
			+ ToString(max) + ") failed with result " + ToString(result));
	}
	return result;
}

GBNumber GBRandomState::InRange(const GBNumber min, const GBNumber max) {
	return GBNumber::MakeRaw(LongInRange(min.data, max.data));
}

float GBRandomState::FloatInRange(const float min, const float max) {
	return min + (float)(unsigned long)GenerateLong() / (float)0xFFFFFFFFUL * (max - min);
}

GBAngle GBRandomState::Angle() {
	return InRange(kEpsilon - kPi, kPi);
}

GBVector GBRandomState::Vector(const GBNumber maxLength) {
	return GBFinePoint::MakePolar(InRange(0, maxLength), Angle());
}

GBColor GBRandomState::Color() {
	return GBColor(FloatInRange(0, 1), FloatInRange(0, 1), FloatInRange(0, 1));
}

// FIXME: ColorNear is broken
GBColor GBRandomState::ColorNear(const GBColor & old, float dist) {
	return GBColor(FloatInRange(old.Red() - dist, old.Red() + dist),
					FloatInRange(old.Green() - dist, old.Green() + dist),
					FloatInRange(old.Blue() - dist, old.Blue() + dist));
}

bool GBRandomState::Boolean(const GBNumber probability) {
	return InRange(0, GBNumber(1) - kEpsilon) < probability;
}

bool GBRandomState::Boolean(const long num, const long denom) {
	return LongInRange(0, denom - 1) < num;
}

long GBRandomState::GetSeed() const {
	return seed;
}

void GBRandomState::SetSeed(const long newSeed) {
	seed = newSeed;
}